src/main/scala/nl/knaw/dans/api/sword2/DepositHandler.scala

```
@@ -263,14 +264,24 @@ object DepositHandler {
       }
     }
 
+    val itemsToResolve = fetchItems diff fetchItemsInBagStore
     for {
-      _ <- resolveFetchItems(bagitDir, fetchItems diff fetchItemsInBagStore)
+      _ <- resolveFetchItems(bagitDir, itemsToResolve)
+      _ <- removeFromFetchTxt(bagitDir, itemsToResolve)
       bag = getBagFromDir(bagitDir)
       validationResult = bag.verifyValid
       _ <- handleValidationResult(bag, validationResult, fetchItemsInBagStore)
     } yield ()
   }
 
+  private def removeFromFetchTxt(bagDir: File, items: Seq[FetchTxt.FilenameSizeUrl]): Try[Unit] = Try {
+    getFetchTxt(bagDir).foreach { case fetchTxt =>
```

I think it is better to use a `map` here:

```scala
getFetchTxt(bagDir)
	.map(fetchTxt => Try {
		items.foreach(fetchTxt.remove)
		if (fetchTxt.isEmpty) new File(bagDir, "fetch.txt").delete()
		getBagFromDir(bagDir).write(new FileSystemWriter(bagFactory), bagDir)
	})
	.map(_ => ())
	.getOrElse(Success(()))
```

---

Why? It looks more complicated. What is the advantage?

---

Seems more obvious to me. In general I'm always careful with the use of `foreach`, as it does not return anything. In the `items.foreach(fetchTxt.remove)` it is a logical choice to do so, but to me it is not at the `getFetchTxt.foreach`. I'd say you want a `Try` back, so you use `map`.

Also, the `getOrElse` makes it more explicitly clear that it is still OK if the `getFetchTxt` returns a `None`. For that to work you also require a `map`.

Sidenote: you could also merge the two `map`s and just return a `()` at the end of the first `map`. The reason I choose for the approach with the second `map` is purely esthetic.

Since each of the three parts of the first `map` are functions that might throw an exception (welcome to the stupidly impure world of Java!), we require this `Try` around it to catch all exceptions and return in `Failure` rather than a `Success` if an exception is thrown.

---

What if the delete fails: Isn't that notified with a boolean result rather than an exception?

---

it's indicated with both: https://docs.oracle.com/javase/8/docs/api/java/io/File.html#delete--

---

The API is a bit vague I think. Surely you cannot do both at the same time. A few tests seem to indicate that deleting file that doesn't exist of for which you have no permissions to delete it are handled by returning `false` rather than throwing an exception, so @jo-pol 's point is valid.